module/executiondatasync/optimistic_sync/execution_result/info_provider.go

```diff
@@ -97,9 +95,43 @@ func (e *Provider) ExecutionResultInfo(
 		}, nil
 	}
 
-	result, executorIDs, err := e.findResultAndExecutors(blockID, criteria)
-	if err != nil {
-		return nil, fmt.Errorf("failed to find result and executors for block ID %v: %w", blockID, err)
+	result, executorIDs, findResultErr := e.findResultAndExecutors(blockID, criteria)
+	if findResultErr != nil {
+		findResultErr = fmt.Errorf("failed to find result and executors for block ID %v: %w", blockID, findResultErr)
+		// We want to return a more specific error when no matching execution results were found.
+		// This helps callers understand that their criteria likely cannot be met — especially if the
+		// block is already sealed.
+
+		// Step 1: Get the block header
+		header, err := e.headers.ByBlockID(blockID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get header by block ID %v: %w", blockID, findResultErr)
+		}
+
+		// Step 2a: Lookup the finalized block ID at this height
+		blockIDFinalized, err := e.headers.BlockIDByHeight(header.Height)
+		if err != nil {
+			return nil, fmt.Errorf("failed to lookup finalized block ID at height %d: %w", header.Height, findResultErr)
+		}
+
+		// Step 2b: Check if this block is finalized.
+		// If BlockIDByHeight returns an ID that doesn't match, not finalized.
+		isFinalized := blockIDFinalized == blockID
+
+		// Step 3: Check sealed status only if block finalized.
+		if isFinalized {
+			sealedHeader, err := e.state.Sealed().Head()
+			if err != nil {
+				return nil, fmt.Errorf("failed to lookup sealed header: %w", err)
+			}
+
+			// If block is sealed, and didn't find any matching results return an error
+			if header.Height <= sealedHeader.Height {
+				return nil, optimistic_sync.NewCriteriaNotMetError(blockID)
+			}
+		}
```

Comment:
consider refactoring this into a helper function (something like `isSealed(blockID)`). that would keep this function's logic better contained